jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/server/ServerCloseTest.java

```diff
@@ -0,0 +1,278 @@
+ */
+public class ServerCloseTest
+{
+    private WebSocketClient client;
+    private Server server;
+    private ServerCloseCreator serverEndpointCreator;
+
+    @BeforeEach
+    public void startServer() throws Exception
+    {
+        server = new Server();
+
+        ServerConnector connector = new ServerConnector(server);
+        connector.setPort(0);
+        server.addConnector(connector);
+
+        ServletContextHandler context = new ServletContextHandler();
+        context.setContextPath("/");
+
+        ServletHolder closeEndpoint = new ServletHolder(new WebSocketServlet()
+        {
+            @Override
+            public void configure(WebSocketServletFactory factory)
+            {
+                factory.setIdleTimeout(Duration.ofSeconds(2));
+                serverEndpointCreator = new ServerCloseCreator(factory);
+                factory.setCreator(serverEndpointCreator);
+            }
+        });
+        context.addServlet(closeEndpoint, "/ws");
+        JettyWebSocketServletContainerInitializer.configure(context);
+
+        HandlerList handlers = new HandlerList();
+        handlers.addHandler(context);
+        handlers.addHandler(new DefaultHandler());
+
+        server.setHandler(handlers);
+
+        server.start();
+    }
+
+    @AfterEach
+    public void stopServer() throws Exception
+    {
+        server.stop();
+    }
+
+    @BeforeEach
+    public void startClient() throws Exception
+    {
+        client = new WebSocketClient();
+        client.setIdleTimeout(Duration.ofSeconds(2));
+        client.start();
+    }
+
+    @AfterEach
+    public void stopClient() throws Exception
+    {
+        client.stop();
+    }
+
+    private void close(Session session)
+    {
+        if (session != null)
+        {
+            session.close();
+        }
+    }
+
+    /**
+     * Test fast close (bug #403817)
+     *
+     * @throws Exception on test failure
+     */
+    @Test
+    public void fastClose() throws Exception
+    {
+        ClientUpgradeRequest request = new ClientUpgradeRequest();
+        request.setSubProtocols("fastclose");
+        CloseTrackingEndpoint clientEndpoint = new CloseTrackingEndpoint();
+
+        URI wsUri = WSURI.toWebsocket(server.getURI().resolve("/ws"));
+        Future<Session> futSession = client.connect(clientEndpoint, wsUri, request);
+
+        Session session = null;
+        try
+        {
+            session = futSession.get(5, SECONDS);
+
+            // Verify that client got close
+            clientEndpoint.assertReceivedCloseEvent(5000, is(StatusCode.NORMAL), containsString(""));
+
+            // Verify that server socket got close event
+            AbstractCloseEndpoint serverEndpoint = serverEndpointCreator.pollLastCreated();
+            assertThat("Fast Close Latch", serverEndpoint.closeLatch.await(5, SECONDS), is(true));
+            assertThat("Fast Close.statusCode", serverEndpoint.closeStatusCode, is(StatusCode.ABNORMAL));
```

Comment:
In this testcase, during the server `onOpen()` it issued a `session.close()`.
Since the server's `onOpen()` never finished successfully, it the websocket connection was never opened for read on the server side, so it cannot read the response from the client.  Hence the ABNORMAL closure.

Reply:
is closing in onOpen abnormal?   I can imagine services that in onOpen send a few messages and then immediately do a normal close and then return.   I think that is normal!

Reply:
closing in onOpen is abnormal & wasClean==false.
this is because the close handshake cannot be completed.

Reply:
the close handshake can be completed, we send the close frame, we return from onWebSocketConnect the onOpen callback is succeeded, so we can now read frames, we read the close response which is normal status code